OpenRA.Game/Server/VoteKickTracker.cs

```diff
@@ -0,0 +1,178 @@
+				EndKickVote();
+				return;
+			}
+
+			if (voteKickTimer.ElapsedMilliseconds > server.Settings.VoteKickTimer)
+			{
+				server.SendLocalizedMessage(VoteKickEnded, Translation.Arguments("kickee", kickee.Client.Name));
+				EndKickVoteAndBlockKicker();
+			}
+		}
+
+		public bool VoteKick(Connection conn, Session.Client kicker, Connection kickeeConn, Session.Client kickee, int kickeeID, bool vote)
+		{
+			if (server.State != ServerState.GameStarted
+				|| (kickee.IsAdmin && server.Type != ServerType.Dedicated)
+				|| (!vote && this.kickee.Client == null) // Disallow starting a vote with a downvote
+				|| (this.kickee.Client != null && this.kickee.Client != kickee) // Disallow starting new votes when one is already ongoing.
+				|| !ClientHasPower(kicker))
+			{
+				server.SendLocalizedMessageTo(conn, UnableToStartAVote);
+				return false;
+			}
+
+			// Prevent vote kick spam abuse.
+			if (voteKickTimer == null && failedVoteKickers.TryGetValue(kicker, out var time))
+			{
+				if (time + server.Settings.VoteKickerCooldown > kickeeConn.ConnectionTimer.ElapsedMilliseconds)
+				{
+					server.SendLocalizedMessageTo(conn, UnableToStartAVote);
+					return false;
+				}
+				else
+					failedVoteKickers.Remove(kicker);
+			}
+
+			var eligeablePlayers = server.Conns.Count(conn => ClientHasPower(server.GetClient(conn)));
+
+			// If only 2 players are playing, they should not be able to kick eachother.
+			if (!kickee.IsObserver && eligeablePlayers < 3)
+			{
+				server.SendLocalizedMessageTo(conn, InsufficientVotes, Translation.Arguments("kickee", kickee.Name));
+				return false;
+			}
+
+			if (voteKickTimer == null)
+			{
+				Log.Write("server", $"Vote kick started on {kickeeID}.");
+				voteKickTimer = Stopwatch.StartNew();
+				server.SendLocalizedMessage(VoteKickStarted, Translation.Arguments("kicker", kicker.Name, "kickee", kickee.Name));
+				server.DispatchServerOrdersToClients(new Order("StartKickVote", null, false) { ExtraData = (uint)kickeeID }.Serialize());
+				this.kickee = (kickee, kickeeConn);
+				voteKickerStarter = (kicker, conn);
+			}
+
+			if (!voteTracker.ContainsKey(conn.PlayerIndex))
+				voteTracker[conn.PlayerIndex] = vote;
+			else
+			{
+				server.SendLocalizedMessageTo(conn, AlreadyVoted, null);
+				return false;
+			}
+
+			var votesFor = 0;
+			var votesAgainst = 0;
+			foreach (var c in voteTracker)
+			{
+				if (c.Value)
+					votesFor++;
+				else
+					votesAgainst++;
+			}
+
+			var percentage = votesFor * 100 / eligeablePlayers;
+			server.SendLocalizedMessage(VoteKickProgress, Translation.Arguments("kickee", kickee.Name, "percentage", percentage));
+
+			// If only a single player is playing, allow him to kick anyone.
+			if (percentage > 50 || (kickee.IsObserver && kickee.IsAdmin && eligeablePlayers < 3))
+			{
+				EndKickVote();
+				return true;
+			}
+
+			// As we include the kickee in eligeablePlayers, we need to subtract him from the votes against.
```

Comment:
Why is the kickee eligible? Doesn't that skew the votes "for"?

Reply:
The logic is cleaner this way.

I wanted it to be skewed, this starts affecting it at above 6 players, where instead of needing 3 for, you need 4 for. Though it seems to only affect even numbers.

Reply:
It would feel weird that in a 2v2, you only need 2 votes to kick a player

Reply:
Doesn't this assume that the `kickee` doesn't vote? Where is this guarded against? This will also get weird in a 2v2 where you will want just one other player voting "no kick" to end the voting. (But this calculation would result in `votesAgainst * 100 / eligablePlayers == 1 * 100 / 3 == 33` which is lower than 50.) Can we just start `votesAgainst` as `1` if the kickee can vote?

Reply:
Self-kicking is guarded against in LobbyCommands. It's there as the translation is there

And you're right, votes against needs a better check